Rename ResetPasswordToken to UserToken

kimlisa opened the pull request:
#### Description
- Renamed ResetPasswordToken to a more general UserToken to to help create user token for other purposes ie: recovery tokens
- Setting different `subkind` differentiates a user token from other tokens

kimlisa committed:
Rename ResetPasswordToken to UserToken in api/*

kimlisa committed:
Rename ResetPasswordToken with UserToken in lib/*

kimlisa committed:
Rename in tctl

alex-kovoy reviewed (approved):
the changes look good to me, i would just double check if we do not break backward compatibility with previous version (like older proxies being able to use newer auth service). If this is the case, we may just keep the current proto definitions (for auth methods) for a while.

russjones commented on api/client/proto/authservice.proto:
```diff
@@ -269,16 +269,13 @@ message RequestStateSetter {
-    string TokenID = 1 [ (gogoproto.jsontag) = "token" ];
-}
+// RotateUserTokenSecretsRequest is a request to rotate token secrets.
+message RotateUserTokenSecretsRequest { string TokenID = 1 [ (gogoproto.jsontag) = "token" ]; }
```
Can you test backward compatibility here? I'm not sure if you can rename messages like this without breaking backward compatibility.

kimlisa replied:
@alex-kovoy @russjones I tested with an auth that included the new changes with an older proxy `v6.2.8` in cloud staging, and I tested each endpoints that were affected by the change and it worked as expected. 

The only endpoint I added was `RotateUserTokenSecrets`, and this will call the fallback `RotateResetPasswordTokenSecrets` if former returns not implemented.

The renaming of message name should not matter according to here: https://developers.google.com/protocol-buffers/docs/encoding#structure, I didn't change the field numbers/types and the json tags that would cause breakage.

codingllama replied:
I'm a bit late to this party (saw the email updates), but Lisa is spot on: as long as we are consistent keeping field types, numbers and json tags, the renames are safe to do. JSON marshaling/unmarshaling is usually the weak link, specially for protos that we might persist.

kimlisa committed:
Check for req type after check and set default

kimlisa committed:
Convert grpc error to trace error

codingllama commented on api/types/events/events.proto:
```diff
@@ -1319,7 +1319,7 @@ message OneOf {
         events.AuthAttempt AuthAttempt = 22;
         events.AccessRequestCreate AccessRequestCreate = 23;
-        events.ResetPasswordTokenCreate ResetPasswordTokenCreate = 24;
+        events.UserTokenCreate UserTokenCreate = 24;
```
This is the one proto change that left me a bit uneasy - if an old version of this was marshaled we'd break trying to read it. It might be safer to deprecate the old `ResetPasswordTokenCreate` and add a new field for `UserTokenCreate`.

I'm not sure what the blast radius of this proto might be, so I'll leave it up to the author and other reviewers.

awly replied:
I **don't** believe this is a breaking change.
The only case where it would be breaking is if `OneOf` got serialized to JSON and back. Protobuf serialization should only care about field numbers.

From what I can tell, we store individual event messages in the audit log backend (like `UserLogin`), not the `OneOf` wrapper.

It should be easy to test though - trigger this event on an older version of teleport and then try to view it in the UI on the newer version of teleport.

kimlisa replied:
> It should be easy to test though - trigger this event on an older version of teleport and then try to view it in the UI on the newer version of teleport.

tested this with audit log screen and no errors were produced

codingllama replied:
It was the chance that it might be serialized as JSON that I was concerned about, but it sounds like we're good.

codingllama reviewed (approved):
LGTM for the proto changes - let me know if you'd like me to review the PR in its entirety too.

codingllama commented on api/types/types.proto:
```diff
@@ -836,24 +836,23 @@ message Namespace {
     option (gogoproto.goproto_stringer) = false;
     option (gogoproto.stringer) = false;
 
     // Kind is a resource kind
```
FYI: this is out of scope for your MR, but I think it would be nice to have some examples within the documentation. Ie, what is usually set in the `Kind` field? What about `SubKind`? Is `Version` a plain number, a semantic version, something else?

kimlisa replied:
I will leave theses as is, b/c all resource templates in teleport follow the same templating with `kind`, `subkind`, and `version`, etc.

codingllama commented on lib/services/local/usertoken.go:
```diff
@@ -145,6 +145,6 @@ func (s *IdentityService) UpsertResetPasswordTokenSecrets(ctx context.Context, s
 const (
-	passwordTokensPrefix = "resetpasswordtokens"
-	secretsPrefix        = "secrets"
+	userTokenPrefix = "usertoken"
```
Doesn't this change have the potential to break in-flight reset token operations?

kimlisa replied:
yes, you'll get `invalid token` screen when user clicks on an old link. i thought about this but these tokens are pretty short-lived with defaults, resets are like 1hr to max 48hr, and invites are 8hr to max 24hr. And these tokens are pretty easy to re-generate for user, so I didn't think it'd be too much of an issue to re-create one (and old tokens will get eventually deleted since we set expiration in backend). Do you think this is too much an inconvenience for user?

kimlisa replied:
I could for a while, also check for the old prefix as well

awly replied:
let's add a check for the old prefix when reading from the new prefix returns a `NotFound` error
it should be a small-ish change and it mostly removes this possible point of confusion

kimlisa replied:
revised to place checks for old prefix

codingllama reviewed (commented):
Reviewed non-proto changes too, sorry for the delay.

russjones reviewed (approved):
Bot.

kimlisa committed:
Add checking for old prefix

codingllama commented on lib/services/local/usertoken.go:
```diff
@@ -60,18 +69,31 @@ func (s *IdentityService) DeleteUserToken(ctx context.Context, tokenID string) e
+
+	// DELETE IN 9.0.0 also delete any tokens with old prefix.
+	startKey = backend.Key(passwordTokensPrefix, tokenID)
+	return s.DeleteRange(ctx, startKey, backend.RangeEnd(startKey))
```
Dropped a trace.Wrap?

kimlisa replied:
revised

codingllama commented on lib/services/local/usertoken.go:
```diff
@@ -145,6 +175,8 @@ func (s *IdentityService) UpsertUserTokenSecrets(ctx context.Context, secrets ty
-	userTokenPrefix = "usertoken"
-	secretsPrefix   = "secrets"
+	// DELETE IN 9.0.0 in favor of userTokenPrefix.
+	passwordTokensPrefix = "resetpasswordtokens"
```
Suggestion: rename to `legacyPasswordTokensPrefix`?

kimlisa replied:
revised

codingllama commented on lib/auth/usertoken.go:
```diff
@@ -247,6 +247,7 @@ func (s *Server) RotateUserTokenSecrets(ctx context.Context, tokenID string) (ty
 	}
 	secrets.SetOTPKey(key.Secret())
 	secrets.SetQRCode(otpQRBuf.Bytes())
+	secrets.SetExpiry(token.Expiry())
```
Should we get test coverage for the missing expiry, so we can avoid regressions in this case?

kimlisa replied:
added test for user token secret creation

codingllama reviewed (changes requested):
LGTM, but we are now missing test coverage for the newly-added logic. I think we should cover at least the read methods to make sure the fallback works.

kimlisa committed:
Address code reviews
- add test for fallback logic
- add test for secrets and token creation settings

codingllama commented on lib/services/local/usertoken.go:
```diff
@@ -0,0 +1,182 @@
+
+// GetUserToken returns a token by its ID.
+func (s *IdentityService) GetUserToken(ctx context.Context, tokenID string) (types.UserToken, error) {
+	item, err := s.Get(ctx, backend.Key(userTokenPrefix, tokenID, paramsPrefix))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			// DELETE IN 9.0.0: fallback for old prefix
+			item, err = s.Get(ctx, backend.Key(LegacyPasswordTokensPrefix, tokenID, paramsPrefix))
+			if err != nil {
+				if trace.IsNotFound(err) {
+					return nil, trace.NotFound("user token(%v) not found", tokenID)
+				}
+				return nil, trace.Wrap(err)
+			}
+		} else {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	token, err := services.UnmarshalUserToken(item.Value)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Suggestion: cut nesting / simplify error handling.  Ditto for all similar blocks in this file.

```suggestion
	item, err := s.Get(ctx, backend.Key(userTokenPrefix, tokenID, paramsPrefix))
	// DELETE IN 9.0.0: fallback for old prefix first.
	if trace.IsNotFound(err) {
		item, err = s.Get(ctx, backend.Key(LegacyPasswordTokensPrefix, tokenID, paramsPrefix))
	}
	// Handle errors from either Get.
	switch {
	case trace.IsNotFound(err):
		return nil, trace.NotFound("user token(%v) not found", tokenID)
	case err != nil:
		return nil, trace.Wrap(err)
	}
```

kimlisa replied:
oh this is much better, applied!

codingllama reviewed (approved):
LGTM.

Thanks for taking the time to refactor and improve the codebase, @kimlisa. Feel free to close my remaining comments yourself.

kimlisa committed:
Apply suggestions, rid of nesting
